source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -0,0 +1,426 @@
+         scheme_header->value().getStringView() == Http::Headers::get().SchemeValues.Http)) {
+      scheme = Http::Headers::get().SchemeValues.Http;
+    }
+
+    const std::string base_path = absl::StrCat(scheme, "://", host_);
+    const std::string callback_path = base_path + config_->callbackPath();
+    const std::string state_path = absl::StrCat(base_path, headers.Path()->value().getStringView());
+
+    const std::string escaped_redirect_uri =
+        Http::Utility::PercentEncoding::encode(callback_path, ":/=&?");
+    const std::string escaped_state = Http::Utility::PercentEncoding::encode(state_path, ":/=&?");
+
+    const std::string new_url =
+        fmt::format(AuthClusterUriParameters, config_->oauthServerHostname(), config_->clientId(),
+                    escaped_redirect_uri, escaped_state);
+    response_headers->setLocation(new_url);
+    decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+    config_->stats().oauth_unauthenticated_rq_.inc();
+
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  // At this point, we *are* on /_oauth. We believe this request comes from the authorization
+  // server and we expect the query strings to contain the information required to get the access
+  // token
+  const auto query_parameters = Http::Utility::parseQueryString(path_str);
+  if (query_parameters.find(queryParamsError()) != query_parameters.end()) {
+    sendUnauthorizedResponse();
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  // if the data we need is not present on the URL, stop execution
+  if (query_parameters.find(queryParamsCode()) == query_parameters.end() ||
+      query_parameters.find(queryParamsState()) == query_parameters.end()) {
+    sendUnauthorizedResponse();
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+
+  auth_code_ = query_parameters.at(queryParamsCode());
+  state_ = Http::Utility::PercentEncoding::decode(query_parameters.at(queryParamsState()));
+
+  Http::Utility::Url state_url;
+  if (!state_url.initialize(state_, false)) {
+    sendUnauthorizedResponse();
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  }
+  const auto scheme = state_url.scheme();
+
+  const std::string cb_url = absl::StrCat(scheme, "://", host_, config_->callbackPath());
+  oauth_client_->asyncGetAccessToken(auth_code_, config_->clientId(), config_->clientSecret(),
+                                     cb_url);
+
+  // pause while we await the next step from the OAuth server
+  return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+}
+
+// Set the legacy Oauth headers.
+void OAuth2Filter::setXForwardedOauthHeaders(Http::RequestHeaderMap& headers,
+                                             const std::string& token) {
+  // Add the x-forwarded headers after inspecting the corresponding cookie values.
+  headers.setInline(authorization_handle.handle(), absl::StrCat("Bearer ", token));
```

Comment:
we're setting the `Authorization` header, but the above comment and function name imply an `x-forwarded` variant? I think that's a bit confusing.